File: backend/app/schemas/ui_schema.py
"""
Pydantic models for UI schema representation.

These models define the structure of the JSON schema sent to the frontend
for form rendering.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field


class QualifierSchema(BaseModel):
    """Schema for a Qualifier on a SubmodelElement."""

    type: str = Field(..., alias="type")
    value: str | int | float | bool | None = None
    valueType: str | None = None
    semanticId: str | None = None
    kind: str | None = None

    class Config:
        populate_by_name = True


class ConstraintSchema(BaseModel):
    """Schema for numeric constraints."""

    min: int | float | None = None
    max: int | float | None = None


class ElementSchema(BaseModel):
    """
    Schema for a SubmodelElement.

    This is the recursive structure that represents any AAS element
    in a form-renderable format.
    """

    idShort: str
    modelType: str
    semanticId: str | None = None
    semanticLabel: str | None = None
    description: dict[str, str] | None = None
    qualifiers: list[QualifierSchema] = Field(default_factory=list)
    cardinality: str = "[1]"
    category: str | None = None

    # Property-specific fields
    valueType: str | None = None
    value: Any = None
    inputType: str | None = None
    step: str | None = None
    constraints: ConstraintSchema | None = None
    unit: str | None = None
    valueId: str | None = None

    # Collection-specific fields
    elements: list["ElementSchema"] | None = None

    # List-specific fields
    typeValueListElement: str | None = None
    orderRelevant: bool | None = None
    valueTypeListElement: str | None = None
    semanticIdListElement: str | None = None
    itemTemplate: "ElementSchema | None" = None
    items: list["ElementSchema"] | None = None

    # MultiLanguageProperty-specific fields
    supportedLanguages: list[str] | None = None

    # File-specific fields
    contentType: str | None = None

    # Range-specific fields
    min: Any = None
    max: Any = None

    # Entity-specific fields
    entityType: str | None = None
    globalAssetId: str | None = None
    specificAssetIds: list[dict[str, str]] | None = None
    statements: list["ElementSchema"] | None = None

    # Relationship-specific fields
    first: str | None = None
    second: str | None = None
    annotations: list["ElementSchema"] | None = None

    # Operation-specific fields
    inputVariables: list["ElementSchema"] | None = None
    outputVariables: list["ElementSchema"] | None = None
    inoutputVariables: list["ElementSchema"] | None = None

    # Event-specific fields
    observed: str | None = None
    direction: str | None = None
    state: str | None = None
    messageTopic: str | None = None
    messageBroker: str | None = None
    lastUpdate: str | None = None
    minInterval: str | None = None
    maxInterval: str | None = None

    class Config:
        extra = "allow"


# Enable forward references
ElementSchema.model_rebuild()


class AdministrationSchema(BaseModel):
    """Schema for AdministrativeInformation."""

    version: str | None = None
    revision: str | None = None
    creator: str | None = None
    templateId: str | None = None


class SubmodelUISchema(BaseModel):
    """
    Complete UI schema for a Submodel.

    This is the top-level structure returned by the parser.
    """

    templateName: str | None = None
    templatePath: str | None = None
    submodelId: str
    idShort: str
    semanticId: str | None = None
    description: dict[str, str] | None = None
    administration: AdministrationSchema | None = None
    elements: list[ElementSchema]
    supplementaryFiles: list[str] = Field(default_factory=list)


class TemplateInfo(BaseModel):
    """Information about an available template."""

    name: str
    path: str
    url: str
    idta_number: str | None = None
    title: str | None = None
    sha: str | None = None


class TemplateVersionInfo(BaseModel):
    """Information about a template version."""

    version: str
    path: str
    sha: str | None = None


class TemplateListResponse(BaseModel):
    """Response for template listing endpoint."""

    templates: list[TemplateInfo]
    total: int
    cached: bool = False


class HealthResponse(BaseModel):
    """Health check response."""

    status: Literal["healthy", "unhealthy"]
    version: str
    details: dict[str, Any] | None = None

File: frontend/src/types/ui-schema.ts
/**
 * TypeScript interfaces for UI schema.
 *
 * These interfaces match the backend Pydantic models and define
 * the structure of data exchanged between frontend and backend.
 */

/**
 * Supported SubmodelElement types.
 */
export type ElementModelType =
  | 'Property'
  | 'SubmodelElementCollection'
  | 'SubmodelElementList'
  | 'MultiLanguageProperty'
  | 'File'
  | 'Blob'
  | 'Range'
  | 'ReferenceElement'
  | 'Entity'
  | 'RelationshipElement'
  | 'AnnotatedRelationshipElement'
  | 'Operation'
  | 'Capability'
  | 'BasicEventElement';

/**
 * Qualifier attached to an element.
 */
export interface QualifierSchema {
  type: string;
  value: string | null;
  valueType: string | null;
  semanticId: string | null;
  kind: string | null;
}

/**
 * Numeric constraints for Property elements.
 */
export interface ConstraintSchema {
  min: number | null;
  max: number | null;
}

/**
 * Schema for a single SubmodelElement.
 *
 * This is a recursive structure - collections and lists contain
 * nested ElementSchema objects.
 */
export interface UIElementSchema {
  // Common fields
  idShort: string;
  modelType: ElementModelType;
  semanticId: string | null;
  semanticLabel: string | null;
  description: Record<string, string> | null;
  qualifiers: QualifierSchema[];
  cardinality: string;
  category: string | null;

  // Property-specific
  valueType?: string;
  value?: unknown;
  inputType?: string;
  step?: string | null;
  constraints?: ConstraintSchema | null;
  unit?: string | null;
  valueId?: string | null;

  // Collection-specific
  elements?: UIElementSchema[];

  // List-specific
  typeValueListElement?: string;
  orderRelevant?: boolean;
  valueTypeListElement?: string | null;
  semanticIdListElement?: string | null;
  itemTemplate?: UIElementSchema | null;
  items?: UIElementSchema[];

  // MultiLanguageProperty-specific
  supportedLanguages?: string[];

  // File-specific
  contentType?: string;

  // Range-specific
  min?: unknown;
  max?: unknown;

  // Entity-specific
  entityType?: string;
  globalAssetId?: string | null;
  specificAssetIds?: Array<{ name: string; value: string }>;
  statements?: UIElementSchema[];

  // Relationship-specific
  first?: string | null;
  second?: string | null;
  annotations?: UIElementSchema[];

  // Operation-specific
  inputVariables?: UIElementSchema[];
  outputVariables?: UIElementSchema[];
  inoutputVariables?: UIElementSchema[];

  // Event-specific
  observed?: string | null;
  direction?: string;
  state?: string;
  messageTopic?: string | null;
  messageBroker?: string | null;
  lastUpdate?: string | null;
  minInterval?: string | null;
  maxInterval?: string | null;
}

/**
 * Administrative information for the submodel.
 */
export interface AdministrationSchema {
  version: string | null;
  revision: string | null;
  creator: string | null;
  templateId: string | null;
}

/**
 * Complete UI schema for a Submodel.
 */
export interface SubmodelUISchema {
  templateName?: string | null;
  templatePath?: string | null;
  submodelId: string;
  idShort: string;
  semanticId: string | null;
  description: Record<string, string> | null;
  administration: AdministrationSchema | null;
  elements: UIElementSchema[];
  supplementaryFiles: string[];
}

/**
 * Information about an available template.
 */
export interface TemplateInfo {
  name: string;
  path: string;
  url: string;
  idta_number: string | null;
  title: string | null;
  sha: string | null;
}

/**
 * Template version information.
 */
export interface TemplateVersionInfo {
  version: string;
  path: string;
  sha: string | null;
}

/**
 * Response from template listing endpoint.
 */
export interface TemplateListResponse {
  templates: TemplateInfo[];
  total: number;
  cached: boolean;
}

/**
 * Validation error detail.
 */
export interface ValidationError {
  field: string;
  message: string;
  code: string | null;
}

/**
 * Form validation result.
 */
export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
  warnings: ValidationError[];
}
